Add global ErrorBoundary component for handling errors in react admin

chihsuan opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Implement a global `ErrorBoundary` component to handle uncaught errors in react admin.

This component will catch any uncaught/unexpected errors during rendering in the React admin app and display a generic error message to the user. This will prevent the app from crashing and provide a better user experience. The error message will show a link to the community forum where users can report the error.

I have added a basic design and copy for the error message. I think we will also need to get some design & copy feedback (cc @verofasulo). Just wanted to get the basic functionality in place and get code review feedback first.

An uncaught error happens in a full-screen page such as core profiler:

Before (white screen):

<img width="1351" alt="Screenshot 2024-06-07 at 15 40 18" src="https://github.com/woocommerce/woocommerce/assets/4344253/47a7fad1-a210-475e-a697-e60fba3ebe64">

After:
![Screenshot 2024-06-10 at 15 49 44](https://github.com/woocommerce/woocommerce/assets/4344253/d1c68fa7-9145-4cd7-8e8b-449fbc7b0208)



An uncaught error happens on home screen:

Before:

<img width="1352" alt="Screenshot 2024-06-07 at 15 43 01" src="https://github.com/woocommerce/woocommerce/assets/4344253/e4664054-f374-46aa-9f90-66774718a3c8">


After:

![Screenshot 2024-06-10 at 15 51 18](https://github.com/woocommerce/woocommerce/assets/4344253/4dcb57c2-c53b-4e1c-b7b8-d17182cf6dab)

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

We need to modify the code to throw an error to test the ErrorBoundary component.

1. Add the following code to the `plugins/woocommerce-admin/client/core-profiler/index.tsx` in the CoreProfilerController component to throw an error:

```js
throw new Error( 'test' );
```

2. Go to the Core Profiler and you should see the error message displayed.
3. Click on the link to community forums and verify that it opens the correct URL
4. Click the reload page button and confirm the page reloads
5. Add the exception code `plugins/woocommerce-admin/client/homescreen/index.tsx`
6. Go to the Home Screen and you should see the error message displayed.


<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

chihsuan committed:
feat: Add ErrorBoundary component for handling errors in WooCommerce admin client

chihsuan committed:
Add changelog

chihsuan committed:
Add tests for error-boundary

rjchow commented on plugins/woocommerce-admin/client/error-boundary/index.tsx:
```diff
@@ -0,0 +1,136 @@
+			return (
+				<div className="woocommerce-error-boundary">
+					<h1 className="woocommerce-error-boundary__heading">
+						{ __( 'Oops, Something went wrong', 'woocommerce' ) }
```
```suggestion
						{ __( 'Oops, something went wrong', 'woocommerce' ) }
```

rjchow commented on plugins/woocommerce-admin/client/error-boundary/index.tsx:
```diff
@@ -0,0 +1,136 @@
+						<Button
+							variant="secondary"
+							onClick={ this.handleOpenIssue }
+							style={ { margin: '10px', padding: '10px 20px' } }
```
Just wondering why inline styles are used here?

chihsuan replied:
My bad, it's was added by ChatGPT 🙈

rjchow commented on plugins/woocommerce-admin/client/error-boundary/index.tsx:
```diff
@@ -0,0 +1,136 @@
+						</Button>
+					</div>
+					<details className="woocommerce-error-boundary__details">
+						<summary>Click for error details</summary>
```
Should we wrap this in `__()` too?

chihsuan replied:
Yes, good catch!

rjchow reviewed (approved):
Looks good, tests well! Thanks for the tests!
Left a few nitpicks

chihsuan committed:
Update plugins/woocommerce-admin/client/error-boundary/index.tsx

Co-authored-by: RJ <27843274+rjchow@users.noreply.github.com>

chihsuan committed:
Wrap text in __()

chihsuan committed:
Address feedback

chihsuan committed:
Fix tests

chihsuan committed:
Improve mobile view

chihsuan committed:
Update button

chihsuan commented:
> Left a few nitpicks

I've addressed feedback and changed the Gtihub button to the Community Support button (pdibGW-3gG-p2#comment-2509).

![Screenshot 2024-06-10 at 15 47 31](https://github.com/woocommerce/woocommerce/assets/4344253/189a765d-5e16-43fa-a46e-81098554e5c8)

moon0326 reviewed (approved):
I love this change! Nice work 👍 

LGTM and tested well 👍 🚀
